### ISSUE TITLE
Allow to skip dependencies (`--no-deps`)

### ISSUE THREAD

**Issue by blueyed** [jazzband/pip-tools#817]
It would be useful if you could specify `--no-deps` for some dependencies, where you do not want pip-compile to include the dependencies, but control it manually.

pip's own options are handled via https://github.com/pypa/pip/blob/f8732ac06dcb7c6af81dcdc0cc3158196b796baf/src/pip/_internal/req/req_file.py#L283 (e.g. `--hash`).

```
echo '--no-deps channels' | pip-compile -q - -o -
```

This might also make sense as a global option to `pip-compile`, but usually is more important for single deps only (i.e. in a line in the `requirements.in` file).

**Comment by graingert** [comment]
I think this is close to https://github.com/jazzband/pip-tools/issues/215

**Comment by blueyed** [comment]
For me `--no-deps` is useful since e.g. "channels" pulls in "daphne", which then pulls in twisted and much more (https://github.com/django/channels/issues/1285), which are really optional deps.

**Comment by atugushev** [comment]
FTR, `--no-deps` is not supported option for `pip install -r requiremtns.txt`. See:

```
$ cat requirements.txt
--no-deps django
django

$ pip install -r requirements.txt
Usage: pip [options]

ERROR: Invalid requirement: --no-deps django
pip: error: no such option: --no-deps
```

**Comment by blueyed** [comment]
Yeah, it should not end up there, but get consumed / used by pip-compile.

**Comment by atugushev** [comment]
The `--no-deps` is global option. So what's the point? It's not gonna work if there are N packages in `reqs.in` with deps, because it'll be all installed without deps.

**Comment by blueyed** [comment]
I've thought about pip-compile using this internally only.
I do not know the internals, but I've thought that pip-compile could resolve those with the "--no-deps" option then, and merging them later - assuming that all reqs are resolved in one batch currently.
Pip's resolver understands the option: https://github.com/pypa/pip/blob/f8732ac06dcb7c6af81dcdc0cc3158196b796baf/src/pip/_internal/resolve.py#L41-L55

**Comment by atugushev** [comment]
> Pip's resolver understands the option: https://github.com/pypa/pip/blob/f8732ac06dcb7c6af81dcdc0cc3158196b796baf/src/pip/_internal/resolve.py#L41-L55

Looks like `ignore_dependencies` works globally.

**Comment by AndydeCleyre** [comment]
Noting here that `--no-deps` as a globally applied option to `pip-sync` (not `pip-compile`) is implemented in [this commit](https://github.com/jazzband/pip-tools/pull/907/commits/ecb4e79181a41151f7966187270608dab7353df4) as part of pull request #907 as a simple pass-through option to `pip`.

**Comment by blueyed** [comment]
@atugushev 
`ignore_dependencies` is an option to the resolver.
Could another resolver be used here?  https://github.com/jazzband/pip-tools/blob/0da08e52294d944ccf59570f2d3397badc41b7d4/piptools/repositories/pypi.py#L221
(for all requirements that have a `--no-deps` option (similar to the existing `-e` option that can be specified with requirements.in entries))

**Comment by AndydeCleyre** [comment]
Honestly I think this is just a problem with particular packages specifying hard deps instead of using extras groupings.

**Comment by AndydeCleyre** [comment]
I don't think this particular workaround (which I *think* achieves the desired behavior) has been articulated here yet:

`requirements.in`:
```ini
-r reqs-with-deps.txt
req-without-dep1
req-without-dep2
```

`reqs-with-deps.in`:
```ini
req1
req2
```

```console
$ pip-compile reqs-with-deps.in && pip-compile --pip-args --no-deps requirements.in
```

**Comment by AndydeCleyre** [comment]
@UncleGoogle 

You're right, I'm sorry!

---

I don't work much with private indexes, so forgive my ignorance here, but why do you need to avoid resolving the dependencies? Why not compile them all, passing `--index-url=PRIVATE --extra-index-url=PYPI` to `pip-compile`? Maybe that doesn't work as I expect.

---

I'll mention that nowadays `uv` can generate TOML lockfiles, which you could grab data from. Here's a hacky example of grabbing just `requests` with its version and hash from a `uv.lock`, using tools I like (Zsh, NestedTextTo, yamlpath, wheezy.template), but you'd probably choose some different ones:

```console
$ wheezy.template =(<<<'@require(version, wheels)
requests==@version\
@for wheel in wheels:
 --hash=@wheel["hash"]
@end') =(toml2nt uv.lock | nt2json | yaml-get -p 'package.[name=requests]')
```

yielding:

```ini
requests==2.32.3 --hash=sha256:70761cfe03c773ceb22aa2f671b4757976145175cdfca038c02654d061d6dcc6
```
